### ISSUE TITLE
Joins generated by ActiveRecord are not compatible with Postgres under certain conditions

### ISSUE THREAD

**Issue by aldhsu** [rails/rails#29104]
### Steps to reproduce
```
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  # Activate the gem you are reporting the issue against.
  gem "activerecord", "5.1.0"
  gem "pg"
  gem "pry"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "postgresql", database: "pg_activerecord_invalid")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :clubs, force: true do |t|
  end

  create_table :club_roles, force: true do |t|
    t.integer :club_id
    t.integer :profile_id
    t.string :type
  end

  create_table :profiles, force: true do |t|
  end

  create_table :parties, force: true do |t|
    t.integer :member_id
  end
end

class Club < ActiveRecord::Base
  has_many :members, dependent: :destroy
  has_many :parties, through: :members
end

class Party < ActiveRecord::Base
  belongs_to :member
end

class ClubRole < ActiveRecord::Base
  belongs_to :club
  belongs_to :profile
end

class Member < ClubRole
  has_many :parties
  belongs_to :club
  belongs_to :profile
end

class Profile < ActiveRecord::Base
  has_many :club_roles
  has_many :members, dependent: :destroy
end

class BugTest < Minitest::Test
  def test_association_stuff
    error = nil

    begin
      club = Club.create!
      club.parties.includes(member: :profile).empty?
    rescue => e
      error = e
    end

    assert_nil error
  end
end
```

### Expected behavior
Joins should only appear in order. In this case
```sql
INNER JOIN club_roles ...
LEFT OUTER JOIN profile ... 
```

### Actual behavior
Postgres needs joins to appear in the order they are used. A join statement cannot use a table that has not yet been joined. 
The SQL produced is:
```sql
...
LEFT OUTER JOIN "club_roles" "members_parties" ON "members_parties"."id" = "parties"."member_id" AND "members_parties"."type" IN ('Member')
LEFT OUTER JOIN "profiles" ON "profiles"."id" = "club_roles"."profile_id" #offending line
INNER JOIN "club_roles" ON "parties"."member_id" = "club_roles"."id" #this should appear before
...
```
Apparently this isn't a limitation for SQLite3, the statement seems valid. I don't have much experience with SQLite3 so I could be wrong.
This seems to be a regression. Switching to 5.0.3 ActiveRecord solves the problem.

### System configuration
postgresql: 9.5.6
pg: 0.18.4
activerecord: 5.1.0

**Comment by alexcameron89** [comment]
@aldhsu have you tried this on 5.1.1? I am able to reproduce your issue on 5.1.0, but it looks like your issue was fixed by 07a611e02351b32100f19495a34ed46d01f195c1 which is included in 5.1.1.

**Comment by aldhsu** [comment]
@alexcameron89 Ah yes it is fixed.
